### PR TITLE
[NeMo-UX] Add option to drop optimizer states

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -320,6 +322,30 @@ def _ema_callback(self, trainer: 'pytorch_lightning.Trainer'):
...
+            )
+            print(f'dropping {checkpoint_path}')
+
+            ## all odd files contain the optimizer states
+            distckpt_files = [f for f in os.listdir(checkpoint_path) if f.endswith("distcp")]
+            if is_global_rank_zero():
+                for f in distckpt_files:
+                    idx = int(f.strip(".distcp").split('_')[-1])
+                    if idx % 2 == 1:
+                        os.remove(checkpoint_path / f)
```
[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py:346]: The implementation of `remove_sharded_tensors` should be sth like:
1. Load metadata (see [here](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/dist_checkpointing/strategies/torch.py#L819))
2. Check if there are any files that start with that separation_hint (otherwise raise error?) and remove them
3. Remove all entries from the metadata that start with a given prefix
    1. metadata consists of 3 parts: `state_dict_metadata`, `planner_data`, `storage_data` - you should remove all entries from all parts
4. Resave the metadata
   1. See [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/filesystem.py#L582)

[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py:346]: For consistency reasons, I would suggest the following order:
1. Save tmp metadata
2. Rename orig (main) metadata to bck
3. Rename tmp metadata to orig (main)
4. Remove .distcp files
5. Remove bck orig metadata

File: nemo/lightning/io/pl.py [hunk 2]
```diff
@@ -243,6 +243,8 @@ def _determine_dist_ckpt_save_strategy(self):
             raise ValueError('Async dist-ckpt save supported only for torch_dist format')
 
         torch_dist_kwargs = {} if self.torch_dist_multiproc is None else dict(thread_count=self.torch_dist_multiproc)
+        ## TODO: make configurable
+        torch_dist_kwargs['separation_hint'] = "optim"
```
[the remark this answers is not in the document]
Comment (ashors1) [reply nemo/lightning/io/pl.py]: No reason. I changed it to `optimizer`, but note that I'll clean this up and make the hint configurable once we nail down the design

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashors1) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py]: Thanks for the detailed suggestions. I've made some changes to this PR and the corresponding megatron one to address this, though this part I'm not super clear on:
>metadata consists of 3 parts: `state_dict_metadata`, `planner_data`, `storage_data` - you should remove all entries from all parts

Could you take a look at the megatron MR and let me know what I am missing?

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -320,6 +322,31 @@ def _ema_callback(self, trainer: 'pytorch_lightning.Trainer'):
...
+        checkpoint_index = len(checkpoints) - self.save_last_n_optim_states - 1
+        if len(checkpoints) > self.save_last_n_optim_states:
+            checkpoint_path = Path(checkpoints[checkpoint_index])
+            checkpoint_path = (
+                checkpoint_path / "weights" if os.path.isdir(checkpoint_path / "weights") else checkpoint_path
+            )
+            ## TODO: verify this
+            if is_global_rank_zero():
+                checkpoint_path = Path(checkpoints[checkpoint_index])
+                checkpoint_path = (
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/callbacks/model_checkpoint.py:341]: is there a reason for repeating this? (it's 3 lines above as well)

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -498,6 +525,9 @@ def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str)
             else:
                 finalize_fn()
 
+            if self.save_last_n_optim_states >= 0 and '-last' in filepath:
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/callbacks/model_checkpoint.py:528]: I'm wondering (not sure) if this shouldn't be inside the `finalize_fn`? So called only after current checkpoint was actually finalized

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py]: explained in MLM MR

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 4 above)
Comment (ashors1) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py]: I think it makes sense to add it to the finalize function so that we always ensure we have at least `n` checkpoints with optimizer states. I made this change.

**Review by dimapihtar — APPROVED** [review]
LGTM, Thank you!

File: nemo/lightning/io/pl.py [hunk 5]
```diff
@@ -302,6 +306,11 @@ def save_sharded_strategy(self) -> 'SaveShardedStrategy':
             self._save_sharded_strategy = self._determine_dist_ckpt_save_strategy()
         return self._save_sharded_strategy
 
+    def drop_optimizer_states(self, path):
+        from megatron.core import dist_checkpointing
```
Comment (mikolajblaz) [inline nemo/lightning/io/pl.py:310]: why is this hidden here? I thought we always import from MCore at the beginning of the file.

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -92,6 +94,9 @@ def __init__(
         self.deferred_ckpts_to_remove: List[List[str]] = []
         self.ckpts_to_link: Dict[str, str] = {}
 
+        # used to drop optimizer states
+        self.base_checkpoint_io = None
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/callbacks/model_checkpoint.py:98]: Is there any reason to keep it as an attribute?
If it's just to avoid recomputing the checkpoint_io unwrapping, I don't think we should store it.

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -516,6 +554,9 @@ def _cb():
             # we don't want to remove the marker until all checkpointing is done.
             self.remove_checkpoint_unfinished_marker(filepath, barrier_before=True)
 
+            if self.save_last_n_optim_states >= 0 and '-last' not in filepath:
+                self._drop_optimizer_states(trainer, filepath, global_step - 1)  ## want step to be zero-indexed here
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/callbacks/model_checkpoint.py:558]: why zero-indexed here?

**Review by mikolajblaz — DISMISSED** [review]
Thanks, just some small comments

File: nemo/lightning/io/pl.py (hunk 5 above)
Comment (ashors1) [reply nemo/lightning/io/pl.py:310]: @marcromeyn might be able to answer better, but the `dist_checkpointing` import is hidden throughout this file. I did the same for consistency

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 7 above)
Comment (ashors1) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py:558]: The step is zero-indexed in the checkpoint name. For example, I see this in a recent log:
```
Async checkpoint save for step 1 (test_logdir/default/v1/checkpoints/...-step=0-....ckpt) finalized successfully. finalized successfully.
```

File: nemo/lightning/io/pl.py (hunk 5 above)
Comment (mikolajblaz) [reply nemo/lightning/io/pl.py:310]: I see. Given more specific imports are in global scope ([example](https://github.com/NVIDIA/NeMo/pull/11089/files/eb030ca78479812354249857db4632dd98e9e375#diff-8b8117bc8b07d1fa52030c25e6a7c7920aafa31780e2665ec5fe63939c11a622R30)), I don't see any point in that anymore

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 7 above)
Comment (mikolajblaz) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py:554]: I'm worried this might lead to inconsistencies difficult to manage. @marcromeyn is there some unified "golden" rule of getting the correct step? Or do we always subtract one (and why)?

My point is, this might work for some specific example, but not generalize

File: nemo/lightning/io/pl.py (hunk 5 above)
Comment (ashors1) [reply nemo/lightning/io/pl.py]: I moved all `dist_checkpointing`-related imports to the top-level. @marcromeyn is this okay with you?

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py (hunk 7 above)
Comment (mikolajblaz) [reply nemo/lightning/pytorch/callbacks/model_checkpoint.py:554]: > As long as a user is using any of the strategies in nemo/lightning/pytorch/strategies, which should be the case, this will work

Yeah this is what I was worried, but I guess this we can assume that.

**Review by mikolajblaz — DISMISSED** [review]

**Review by pablo-garay — DISMISSED** [review]
